File: src/core/zip/read.ts
import type fs from 'node:fs';
import { fail } from '../../shared/errors.ts';

const EOCD_SIG = 0x06054b50;
const EOCD64_SIG = 0x06064b50;
const EOCD64_LOCATOR_SIG = 0x07064b50;
const CENTRAL_SIG = 0x02014b50;
const LOCAL_SIG = 0x04034b50;

const EOCD_MIN = 22;
const MAX_COMMENT = 0xffff;

/** Unix file-type mask and the symlink type, as stored in st_mode. */
const S_IFMT = 0xf000;
const S_IFLNK = 0xa000;

export interface ZipEntry {
  name: string;
  method: number;
  crc32: number;
  compressedSize: number;
  uncompressedSize: number;
  localHeaderOffset: number;
  /** Unix st_mode when the archive records one, else null. */
  unixMode: number | null;
  isDirectory: boolean;
  isSymlink: boolean;
  /** Bit 0 of the general-purpose flags: the entry is encrypted. */
  encrypted: boolean;
}

async function readChunk(handle: fs.promises.FileHandle, length: number, position: number): Promise<Buffer> {
  if (length <= 0) return Buffer.alloc(0);
  const buffer = Buffer.alloc(length);
  const { bytesRead } = await handle.read(buffer, 0, length, position);
  return bytesRead === length ? buffer : buffer.subarray(0, bytesRead);
}

/**
 * Find the End Of Central Directory record by scanning backwards. Its position
 * is not fixed because a ZIP may carry a trailing comment of up to 64 KiB.
 */
async function findEocd(handle: fs.promises.FileHandle, size: number): Promise<{ buffer: Buffer; offset: number }> {
  const window = Math.min(size, EOCD_MIN + MAX_COMMENT);
  const start = size - window;
  const tail = await readChunk(handle, window, start);
  for (let i = tail.length - EOCD_MIN; i >= 0; i -= 1) {
    if (tail.readUInt32LE(i) !== EOCD_SIG) continue;
    const commentLength = tail.readUInt16LE(i + 20);
    // The comment length must account for exactly the remaining bytes, which
    // is what tells a real EOCD apart from those four bytes appearing in data.
    if (i + EOCD_MIN + commentLength === tail.length) {
      return { buffer: tail.subarray(i), offset: start + i };
    }
  }
  return fail('zipInvalid', 'no end-of-central-directory record');
}

interface Directory {
  entryCount: number;
  centralOffset: number;
  centralSize: number;
}

async function readDirectoryLocation(handle: fs.promises.FileHandle, size: number): Promise<Directory> {
  const { buffer: eocd, offset: eocdOffset } = await findEocd(handle, size);
  let entryCount = eocd.readUInt16LE(10);
  let centralSize = eocd.readUInt32LE(12);
  let centralOffset = eocd.readUInt32LE(16);

  // 0xffff / 0xffffffff are the Zip64 sentinels: the real values live in the
  // Zip64 record, located through a 20-byte locator just before the EOCD.
  const needsZip64 =
    entryCount === 0xffff || centralSize === 0xffffffff || centralOffset === 0xffffffff;
  if (needsZip64 && eocdOffset >= 20) {
    const locator = await readChunk(handle, 20, eocdOffset - 20);
    if (locator.length === 20 && locator.readUInt32LE(0) === EOCD64_LOCATOR_SIG) {
      const zip64Offset = Number(locator.readBigUInt64LE(8));
      const zip64 = await readChunk(handle, 56, zip64Offset);
      if (zip64.length >= 56 && zip64.readUInt32LE(0) === EOCD64_SIG) {
        entryCount = Number(zip64.readBigUInt64LE(32));
        centralSize = Number(zip64.readBigUInt64LE(40));
        centralOffset = Number(zip64.readBigUInt64LE(48));
      }
    }
  }

  if (centralOffset < 0 || centralOffset + centralSize > size) {
    return fail('zipInvalid', 'central directory lies outside the file');
  }
  return { entryCount, centralOffset, centralSize };
}

/**
 * Zip64 stores oversized fields in extra-field block 0x0001, present only for
 * the values that actually overflowed - so the block is read positionally in
 * the same order the 32-bit fields appear.
 */
function applyZip64Extra(entry: ZipEntry, extra: Buffer): void {
  let offset = 0;
  while (offset + 4 <= extra.length) {
    const id = extra.readUInt16LE(offset);
    const size = extra.readUInt16LE(offset + 2);
    const body = extra.subarray(offset + 4, offset + 4 + size);
    if (id === 0x0001) {
      let cursor = 0;
      const next = (): number | null => {
        if (cursor + 8 > body.length) return null;
        const value = Number(body.readBigUInt64LE(cursor));
        cursor += 8;
        return value;
      };
      if (entry.uncompressedSize === 0xffffffff) entry.uncompressedSize = next() ?? entry.uncompressedSize;
      if (entry.compressedSize === 0xffffffff) entry.compressedSize = next() ?? entry.compressedSize;
      if (entry.localHeaderOffset === 0xffffffff) entry.localHeaderOffset = next() ?? entry.localHeaderOffset;
      return;
    }
    offset += 4 + size;
  }
}

/** Every entry described by the central directory, which is authoritative. */
export async function readEntries(handle: fs.promises.FileHandle, size: number): Promise<ZipEntry[]> {
  const { entryCount, centralOffset, centralSize } = await readDirectoryLocation(handle, size);
  const central = await readChunk(handle, centralSize, centralOffset);
  const entries: ZipEntry[] = [];
  let offset = 0;

  while (entries.length < entryCount) {
    if (offset + 46 > central.length) break;
    if (central.readUInt32LE(offset) !== CENTRAL_SIG) {
      return fail('zipInvalid', 'bad central directory header', { at: offset });
    }
    const versionMadeBy = central.readUInt16LE(offset + 4);
    const flags = central.readUInt16LE(offset + 8);
    const nameLength = central.readUInt16LE(offset + 28);
    const extraLength = central.readUInt16LE(offset + 30);
    const commentLength = central.readUInt16LE(offset + 32);
    const externalAttributes = central.readUInt32LE(offset + 38);

    const nameStart = offset + 46;
    const nameEnd = nameStart + nameLength;
    if (nameEnd > central.length) return fail('zipInvalid', 'truncated entry name');
    // Bit 11 marks the name as UTF-8. Older archivers use the local codepage;
    // decoding those as UTF-8 is the common convention and is lossless for the
    // ASCII paths every archive we consume actually uses.
    const name = central.subarray(nameStart, nameEnd).toString('utf8');

    // Only the Unix host (3) puts st_mode in the high half of the attributes.
    const unixMode = (versionMadeBy >> 8) === 3 ? (externalAttributes >>> 16) & 0xffff : null;

    const entry: ZipEntry = {
      name,
      method: central.readUInt16LE(offset + 10),
      crc32: central.readUInt32LE(offset + 16),
      compressedSize: central.readUInt32LE(offset + 20),
      uncompressedSize: central.readUInt32LE(offset + 24),
      localHeaderOffset: central.readUInt32LE(offset + 42),
      unixMode,
      isDirectory: name.endsWith('/') || name.endsWith('\\'),
      isSymlink: unixMode !== null && (unixMode & S_IFMT) === S_IFLNK,
      encrypted: (flags & 0x1) !== 0
    };
    applyZip64Extra(entry, central.subarray(nameEnd, nameEnd + extraLength));
    entries.push(entry);
    offset = nameEnd + extraLength + commentLength;
  }

  return entries;
}

/** Byte offset at which an entry's compressed data begins. */
export async function dataOffsetOf(handle: fs.promises.FileHandle, entry: ZipEntry): Promise<number> {
  const header = await readChunk(handle, 30, entry.localHeaderOffset);
  if (header.length < 30 || header.readUInt32LE(0) !== LOCAL_SIG) {
    return fail('zipInvalid', 'bad local file header', { name: entry.name });
  }
  const nameLength = header.readUInt16LE(26);
  const extraLength = header.readUInt16LE(28);
  return entry.localHeaderOffset + 30 + nameLength + extraLength;
}

File: src-tauri/crates/core/src/zip/read.rs
use std::fs::File;
use std::io::{Read, Seek, SeekFrom};

use crate::bytes::Le;
use crate::error::{fail, Code, Result};

const EOCD_SIG: u32 = 0x0605_4b50;
const EOCD64_SIG: u32 = 0x0606_4b50;
const EOCD64_LOCATOR_SIG: u32 = 0x0706_4b50;
const CENTRAL_SIG: u32 = 0x0201_4b50;
const LOCAL_SIG: u32 = 0x0403_4b50;

const EOCD_MIN: usize = 22;
const MAX_COMMENT: usize = 0xffff;

/// Unix file-type mask and the symlink type, as stored in `st_mode`.
const S_IFMT: u16 = 0xf000;
const S_IFLNK: u16 = 0xa000;

#[derive(Debug, Clone)]
pub struct ZipEntry {
    pub name: String,
    pub method: u16,
    pub crc32: u32,
    pub compressed_size: u64,
    pub uncompressed_size: u64,
    pub local_header_offset: u64,
    /// Unix `st_mode` when the archive records one.
    pub unix_mode: Option<u16>,
    pub is_directory: bool,
    pub is_symlink: bool,
    /// Bit 0 of the general-purpose flags.
    pub encrypted: bool,
}

/// The shared `Le` accessors return `Option`; a malformed archive is an error
/// with a code, so they are adapted once here rather than at thirty call
/// sites. The bounds checking itself stays in `crate::bytes`.
trait LeZip {
    fn zu16(&self, offset: usize) -> Result<u16>;
    fn zu32(&self, offset: usize) -> Result<u32>;
    fn zu64(&self, offset: usize) -> Result<u64>;
}

impl LeZip for [u8] {
    // Qualified calls, so these cannot resolve back to themselves.
    fn zu16(&self, offset: usize) -> Result<u16> {
        match Le::u16_at(self, offset) {
            Some(value) => Ok(value),
            None => fail(Code::ZipInvalid, "truncated: expected 2 bytes"),
        }
    }

    fn zu32(&self, offset: usize) -> Result<u32> {
        match Le::u32_at(self, offset) {
            Some(value) => Ok(value),
            None => fail(Code::ZipInvalid, "truncated: expected 4 bytes"),
        }
    }

    fn zu64(&self, offset: usize) -> Result<u64> {
        match Le::u64_at(self, offset) {
            Some(value) => Ok(value),
            None => fail(Code::ZipInvalid, "truncated: expected 8 bytes"),
        }
    }
}

fn read_chunk(file: &mut File, length: usize, position: u64) -> Result<Vec<u8>> {
    if length == 0 {
        return Ok(Vec::new());
    }
    file.seek(SeekFrom::Start(position))
        .map_err(|error| crate::Error::new(Code::ZipInvalid, format!("seek failed: {error}")))?;
    let mut buffer = vec![0u8; length];
    let mut filled = 0;
    while filled < length {
        match file.read(buffer.get_mut(filled..).unwrap_or(&mut [])) {
            Ok(0) => break,
            Ok(read) => filled += read,
            Err(error) if error.kind() == std::io::ErrorKind::Interrupted => {}
            Err(error) => {
                return fail(Code::ZipInvalid, format!("read failed: {error}"));
            }
        }
    }
    buffer.truncate(filled);
    Ok(buffer)
}

/// Locate the End Of Central Directory record by scanning backwards: its
/// position is not fixed, because a ZIP may carry a comment of up to 64 KiB.
fn find_eocd(file: &mut File, size: u64) -> Result<(Vec<u8>, u64)> {
    let window = std::cmp::min(size, (EOCD_MIN + MAX_COMMENT) as u64);
    let start = size - window;
    let tail = read_chunk(file, usize::try_from(window).unwrap_or(0), start)?;
    if tail.len() < EOCD_MIN {
        return fail(Code::ZipInvalid, "file is shorter than an EOCD record");
    }

    for index in (0..=tail.len() - EOCD_MIN).rev() {
        if tail.zu32(index)? != EOCD_SIG {
            continue;
        }
        let comment_length = usize::from(tail.zu16(index + 20)?);
        // The comment length must account for exactly the remaining bytes,
        // which is what distinguishes a real EOCD from those four bytes
        // happening to appear inside compressed data.
        if index + EOCD_MIN + comment_length == tail.len() {
            let record = tail.get(index..).unwrap_or(&[]).to_vec();
            return Ok((record, start + index as u64));
        }
    }
    fail(Code::ZipInvalid, "no end-of-central-directory record")
}

struct Directory {
    entry_count: u64,
    central_offset: u64,
    central_size: u64,
}

fn read_directory_location(file: &mut File, size: u64) -> Result<Directory> {
    let (eocd, eocd_offset) = find_eocd(file, size)?;
    let mut entry_count = u64::from(eocd.zu16(10)?);
    let mut central_size = u64::from(eocd.zu32(12)?);
    let mut central_offset = u64::from(eocd.zu32(16)?);

    // 0xffff / 0xffffffff are the Zip64 sentinels: the real values live in the
    // Zip64 record, reached through a 20-byte locator just before the EOCD.
    let needs_zip64 =
        entry_count == 0xffff || central_size == 0xffff_ffff || central_offset == 0xffff_ffff;
    if needs_zip64 && eocd_offset >= 20 {
        let locator = read_chunk(file, 20, eocd_offset - 20)?;
        if locator.len() == 20 && locator.zu32(0)? == EOCD64_LOCATOR_SIG {
            let zip64_offset = locator.zu64(8)?;
            let zip64 = read_chunk(file, 56, zip64_offset)?;
            if zip64.len() >= 56 && zip64.zu32(0)? == EOCD64_SIG {
                entry_count = zip64.zu64(32)?;
                central_size = zip64.zu64(40)?;
                central_offset = zip64.zu64(48)?;
            }
        }
    }

    if central_offset.saturating_add(central_size) > size {
        return fail(Code::ZipInvalid, "central directory lies outside the file");
    }
    Ok(Directory {
        entry_count,
        central_offset,
        central_size,
    })
}

/// Zip64 stores oversized fields in extra-field block 0x0001, present only for
/// the values that actually overflowed - so the block is read positionally in
/// the same order the 32-bit fields appear.
fn apply_zip64_extra(entry: &mut ZipEntry, extra: &[u8]) -> Result<()> {
    let mut offset = 0usize;
    while offset + 4 <= extra.len() {
        let id = extra.zu16(offset)?;
        let size = usize::from(extra.zu16(offset + 2)?);
        let body = extra.get(offset + 4..offset + 4 + size).unwrap_or(&[]);
        if id == 0x0001 {
            let mut cursor = 0usize;
            let next = |cursor: &mut usize| -> Option<u64> {
                let value = body.zu64(*cursor).ok()?;
                *cursor += 8;
                Some(value)
            };
            if entry.uncompressed_size == 0xffff_ffff {
                if let Some(value) = next(&mut cursor) {
                    entry.uncompressed_size = value;
                }
            }
            if entry.compressed_size == 0xffff_ffff {
                if let Some(value) = next(&mut cursor) {
                    entry.compressed_size = value;
                }
            }
            if entry.local_header_offset == 0xffff_ffff {
                if let Some(value) = next(&mut cursor) {
                    entry.local_header_offset = value;
                }
            }
            return Ok(());
        }
        offset += 4 + size;
    }
    Ok(())
}

/// Every entry described by the central directory, which is authoritative.
pub fn read_entries(file: &mut File, size: u64) -> Result<Vec<ZipEntry>> {
    let location = read_directory_location(file, size)?;
    let central = read_chunk(
        file,
        usize::try_from(location.central_size).unwrap_or(0),
        location.central_offset,
    )?;

    let mut entries: Vec<ZipEntry> = Vec::new();
    let mut offset = 0usize;

    while (entries.len() as u64) < location.entry_count {
        if offset + 46 > central.len() {
            break;
        }
        if central.zu32(offset)? != CENTRAL_SIG {
            return fail(
                Code::ZipInvalid,
                format!("bad central directory header at {offset}"),
            );
        }
        let version_made_by = central.zu16(offset + 4)?;
        let flags = central.zu16(offset + 8)?;
        let name_length = usize::from(central.zu16(offset + 28)?);
        let extra_length = usize::from(central.zu16(offset + 30)?);
        let comment_length = usize::from(central.zu16(offset + 32)?);
        let external_attributes = central.zu32(offset + 38)?;

        let name_start = offset + 46;
        let name_end = name_start + name_length;
        let Some(name_bytes) = central.get(name_start..name_end) else {
            return fail(Code::ZipInvalid, "truncated entry name");
        };
        // Bit 11 marks the name as UTF-8. Older archivers use the local
        // codepage; decoding as UTF-8 is the common convention and is lossless
        // for the ASCII paths every archive we consume actually uses.
        let name = String::from_utf8_lossy(name_bytes).into_owned();

        // Only the Unix host (3) puts st_mode in the high half of the attrs.
        let unix_mode = if version_made_by >> 8 == 3 {
            Some(((external_attributes >> 16) & 0xffff) as u16)
        } else {
            None
        };

        let mut entry = ZipEntry {
            is_directory: name.ends_with('/') || name.ends_with('\\'),
            is_symlink: unix_mode.is_some_and(|mode| mode & S_IFMT == S_IFLNK),
            encrypted: flags & 0x1 != 0,
            name,
            method: central.zu16(offset + 10)?,
            crc32: central.zu32(offset + 16)?,
            compressed_size: u64::from(central.zu32(offset + 20)?),
            uncompressed_size: u64::from(central.zu32(offset + 24)?),
            local_header_offset: u64::from(central.zu32(offset + 42)?),
            unix_mode,
        };
        let extra = central
            .get(name_end..name_end + extra_length)
            .unwrap_or(&[]);
        apply_zip64_extra(&mut entry, extra)?;

        entries.push(entry);
        offset = name_end + extra_length + comment_length;
    }

    Ok(entries)
}

/// Byte offset at which an entry's compressed data begins.
pub fn data_offset_of(file: &mut File, entry: &ZipEntry) -> Result<u64> {
    let header = read_chunk(file, 30, entry.local_header_offset)?;
    if header.len() < 30 || header.zu32(0)? != LOCAL_SIG {
        return fail(
            Code::ZipInvalid,
            format!("bad local file header: {}", entry.name),
        );
    }
    let name_length = u64::from(header.zu16(26)?);
    let extra_length = u64::from(header.zu16(28)?);
    Ok(entry.local_header_offset + 30 + name_length + extra_length)
}
